python-spec/src/somabase/data.py

```diff
@@ -100,3 +191,55 @@ def nnz(self) -> int:
 
     is_sparse: Final = True
     soma_type: Final = "SOMASparseNDArray"
+
+
+#
+# Read types
+#
+
+_T = TypeVar("_T")
+
+
+# Sparse reads are returned as an iterable structure:
+
+
+class ReadIter(Iterator[_T], metaclass=abc.ABCMeta):
+    """SparseRead result iterator allowing users to flatten the iteration."""
+
+    # __iter__ is already implemented as `return self` in Iterator
+    # SOMA implementations must implement __next__.
+
+    # XXX: Considering the name "flat" here too.
+    @abc.abstractmethod
+    def all(self) -> _T:
+        """Returns all the requested data in a single operation."""
+        raise NotImplementedError()
+
+
+class SparseRead(metaclass=abc.ABCMeta):
+    """Intermediate type to allow users to format when reading a sparse array."""
+
+    @abc.abstractmethod
+    def coos(self) -> ReadIter[pyarrow.SparseCOOTensor]:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
```

Comment:
these only make sense if it is a 2D array (matrix).  What is the expected behavior if it is not 2D?  Raise?

Reply:
Raising an error seems like a good idea. Initially this can be a basic `ValueError` but in the future we can build out some more granular exception types to allow better error handling.

Reply:
Sounds good - so the pattern is to have  consistent set of accessor method names, and just error out those that don't make sense. The only downside to this will be documentation - i.e., there may be methods that are nonsensical on a DataFrame but work fine on an NDArray.

Reply:
BTW, given this approach, perhaps this these methods should not be abstract -- i.e., subclasses should inherit the default implementation (raising the "NotImplemented" exception when they are not implemented).

Reply:
un-abstracted these.